### PR TITLE
[FLINK-19264][runtime] Add JobID to ExecutionAttemptID

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Add JobID to ExecutionAttemptID

## Brief change log

Add JobID to ExecutionAttemptID

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionAttemptID.java [hunk 1]
```diff
@@ -33,31 +36,46 @@
...
 	 */
 	public ExecutionAttemptID() {
-		this(new ExecutionVertexID(new JobVertexID(), 0), 0);
+		this(new JobID(), new ExecutionVertexID(new JobVertexID(), 0), 0);
 	}
 
-	public ExecutionAttemptID(ExecutionVertexID executionVertexId, int attemptNumber) {
+	public ExecutionAttemptID(JobID jobId, ExecutionVertexID executionVertexId, int attemptNumber) {
 		Preconditions.checkState(attemptNumber >= 0);
+		this.jobId = jobId;
```
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionAttemptID.java:52]: need to check not null

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this pr @KarmaGYZ 
The change looks good to me except for one minor comment.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing the comments! @KarmaGYZ 
The change looks good to me
Will merge it once AZP gives green

**Comment by zentol** [comment]
Can we also add a test that tries to run 2 identical jobs at the same time?

**Comment by KarmaGYZ** [comment]
> Can we also add a test that tries to run 2 identical jobs at the same time?

Thanks for the comment @zentol . I think it will be good to add a test for it. However, I'm not quite familiar with that logic. What I can find is the `ExecutionGraphConstructionTest`, we could check the `ExecutionAttemptID` for each execution vertex it generates. WDYT?

Otherwise, I think we need to add an e2e test. Or do you have any suggestion? Thanks.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphConstructionTest.java [hunk 2]
```diff
@@ -61,6 +66,36 @@ private ExecutionGraph createExecutionGraph() throws Exception {
...
+
+		ExecutionGraph eg1 = createExecutionGraph();
+		ExecutionGraph eg2 = createExecutionGraph();
+		try {
+			eg1.attachJobGraph(ordered);
+			eg2.attachJobGraph(ordered);
+		}
+		catch (JobException e) {
+			e.printStackTrace();
+			fail("Job failed with exception: " + e.getMessage());
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphConstructionTest.java:93]: I don't see a need for an explicit catch block

**Comment by zentol** [comment]
I originally thought of just using a `MiniCluster` and submitting 2 job graphs directly (as a high-level test), but I suppose the test you added should suffice as well. The actual parallel execution will be covered in FLINK-19123.

**Comment by aljoscha** [comment]
So? Can we merge? 🤔

**Pull request merged**